sql: unsupported comparison during index selection

Regression in `TestLogic -bigtest`.

Assigning to @knz as it seems like you're the last one to change the index selection code - feel free to unassign if this is incorrect.

```
=== RUN   TestLogic/../../../sqllogictest/test/index/between/1/slt_good_0.test
--- new user: root
--- queries start here
panic: unsupported comparison: tuple{int, int, int, int, int} to int [recovered]
        panic: SELECT pk FROM tab1 WHERE (col0 > 9 AND (((((col1 <= 6.38) AND (col0 > 2)) OR (col0 BETWEEN 5 AND 8)) AND (col0 IN (7,6,3,5,0) OR ((((col4 IN (6.36,7.74,1.24))))))))): unsupported comparison: tuple{int, int, int, int, int} to int [recovered]
        panic: SELECT pk FROM tab1 WHERE (col0 > 9 AND (((((col1 <= 6.38) AND (col0 > 2)) OR (col0 BETWEEN 5 AND 8)) AND (col0 IN (7,6,3,5,0) OR ((((col4 IN (6.36,7.74,1.24))))))))): unsupported comparison: tuple{int, int, int, int, int} to int [recovered]
        panic: SELECT pk FROM tab1 WHERE (col0 > 9 AND (((((col1 <= 6.38) AND (col0 > 2)) OR (col0 BETWEEN 5 AND 8)) AND (col0 IN (7,6,3,5,0) OR ((((col4 IN (6.36,7.74,1.24))))))))): unsupported comparison: tuple{int, int, int, int, int} to int

goroutine 227 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4200b61b0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:198 +0x6e
panic(0x53819c0, 0xc420ba0020)
        /usr/local/Cellar/go/1.8/libexec/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc420a8be00, 0x8762080, 0xc4209e1950)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:340 +0x74
panic(0x53819c0, 0xc420ba0020)
        /usr/local/Cellar/go/1.8/libexec/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements.func1(0xc420bd997f, 0xa6)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:484 +0x131
panic(0x5335ce0, 0xc420ba0000)
        /usr/local/Cellar/go/1.8/libexec/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/parser.(*DTuple).Compare(0xc4205f6bc0, 0x5e853c0, 0xc420d61fc0, 0x5e853c0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/parser/datum.go:1505 +0x1e5
github.com/cockroachdb/cockroach/pkg/sql.applyConstraintFlat(0xc4209f4300, 0x5e85820, 0xc4205f6bc0, 0x4, 0x5e853c0, 0xc420d61fc0, 0x0, 0x63b29c9)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/index_selection.go:1723 +0x610
github.com/cockroachdb/cockroach/pkg/sql.applyConstraint(0xc4209f4300, 0xc4209f4480, 0xc420acc510, 0xc4209b3188)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/index_selection.go:1590 +0x51a
github.com/cockroachdb/cockroach/pkg/sql.(*applyConstraintsVisitor).VisitPre(0xc4205f7060, 0x5e7a480, 0xc4209f4300, 0x63b29c9, 0x16, 0xc4209b31f0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/index_selection.go:1519 +0x11f
github.com/cockroachdb/cockroach/pkg/sql/parser.WalkExpr(0x5e6c340, 0xc4205f7060, 0x5e7a480, 0xc4209f4300, 0x5e7a500, 0x63b29c9, 0x5e85301)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/parser/walk.go:557 +0x4c
github.com/cockroachdb/cockroach/pkg/sql/parser.(*OrExpr).Walk(0xc420acc360, 0x5e6c340, 0xc4205f7060, 0xc420acc501, 0x5e7ad40)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/parser/walk.go:371 +0x55
github.com/cockroachdb/cockroach/pkg/sql/parser.WalkExpr(0x5e6c340, 0xc4205f7060, 0x5e7ad40, 0xc420acc360, 0x5e7ad40, 0xc420acc540, 0x4013b01)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/parser/walk.go:560 +0x15a
github.com/cockroachdb/cockroach/pkg/sql/parser.(*AndExpr).Walk(0xc420acc3f0, 0x5e6c340, 0xc4205f7060, 0x5e60301, 0x5e7a140)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/parser/walk.go:49 +0xa7
github.com/cockroachdb/cockroach/pkg/sql/parser.WalkExpr(0x5e6c340, 0xc4205f7060, 0x5e7a140, 0xc420acc3f0, 0x5e7a500, 0x5e6032b, 0x1)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/parser/walk.go:560 +0x15a
github.com/cockroachdb/cockroach/pkg/sql/parser.(*AndExpr).Walk(0xc420acc420, 0x5e6c340, 0xc4205f7060, 0x1, 0x5e7a140)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/parser/walk.go:49 +0xa7
github.com/cockroachdb/cockroach/pkg/sql/parser.WalkExpr(0x5e6c340, 0xc4205f7060, 0x5e7a140, 0xc420acc420, 0x0, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/parser/walk.go:560 +0x15a
github.com/cockroachdb/cockroach/pkg/sql.applyIndexConstraints(0x5e7e760, 0xc420acc420, 0xc4205f6fa0, 0x1, 0x1, 0xc420acc4b0, 0x1)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/index_selection.go:1401 +0x1e6
github.com/cockroachdb/cockroach/pkg/sql.(*planner).selectIndex(0xc420853be0, 0xc420c3e000, 0x0, 0x4010a00, 0xc4209b38b8, 0x4013bfc, 0xc4205f6de0, 0xc420c3e000)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/index_selection.go:214 +0x6b9
github.com/cockroachdb/cockroach/pkg/sql.expandScanNode(0xc420853be0, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0xc420c3e000, 0xc4209b3930, 0x5e7b940, 0xc4209b3a30, 0x49e8ec4)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:279 +0xea
github.com/cockroachdb/cockroach/pkg/sql.doExpandPlan(0xc420853be0, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0x5e82fc0, 0xc420c3e000, 0xc420853be0, 0x5e82c60, 0xc4200b6ab0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:230 +0x665
github.com/cockroachdb/cockroach/pkg/sql.expandRenderNode(0xc420853be0, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0xc420406ea0, 0xc4205768f0, 0x5e7aa00, 0xc4209b3cf0, 0xc4209b3d48)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:290 +0xbd
github.com/cockroachdb/cockroach/pkg/sql.doExpandPlan(0xc420853be0, 0x7fffffffffffffff, 0x0, 0x0, 0x0, 0x5e82f60, 0xc420406ea0, 0xc42065a000, 0x8000000, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:233 +0x2ae8
github.com/cockroachdb/cockroach/pkg/sql.(*planner).expandPlan(0xc420853be0, 0x5e82f60, 0xc420406ea0, 0x5e82f60, 0xc420406ea0, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/expand_plan.go:34 +0x73
github.com/cockroachdb/cockroach/pkg/sql.(*planner).optimizePlan(0xc420853be0, 0x5e82f60, 0xc420406ea0, 0xc420d61ef0, 0x1, 0x1, 0xc420d61e01, 0x5e82f60, 0xc420406ea0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/optimize.go:38 +0xd7
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc420853be0, 0x5e7ae80, 0xc420bf18c0, 0x1, 0xc42001c050, 0xc42001c000, 0xed03fae4c, 0xc422b2d371)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:204 +0x12b
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt(0xc4200f6a80, 0x5e7ae80, 0xc420bf18c0, 0xc420853be0, 0xc4209b0001, 0xed03fae4c, 0x22b2d335, 0x637b2e0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1381 +0xa3
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc4200f6a80, 0x5e7ae80, 0xc420bf18c0, 0xc420853be0, 0x101, 0xc420853ae8, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1167 +0x2bd
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtsInCurrentTxn(0xc4200f6a80, 0xc420d61cc0, 0x1, 0x1, 0xc420853be0, 0xc420853ae8, 0x101, 0x3, 0xc4209f91e0, 0xc420a92700, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:902 +0x7f0
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc4200f6a80, 0xc420853be0, 0x1, 0xc420853ae8, 0xc420d61d20, 0xc420d61cc0, 0x1, 0x1, 0x0, 0x0, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:816 +0xf6
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest.func2(0xc420d6f200, 0xc420d61d20, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:655 +0x1bc
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc420d6f200, 0xc420b10101, 0xc4202d64c0, 0xc420b17570, 0xc400000000, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:511 +0x1ab
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc4200f6a80, 0xc420853a80, 0xc420bd997f, 0xa6, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:677 +0x583
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc4200f6a80, 0xc420853a80, 0xc420bd997f, 0xa6, 0x0, 0x0, 0x0, 0x0, 0xc420bd9900)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:490 +0x12d
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).executeStatements(0xc420a8be00, 0x8762080, 0xc4209e1950, 0xc420bd997f, 0xa6, 0x0, 0x0, 0x0, 0x0, 0x43fdb01, ...)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:813 +0x94
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc420a8be00, 0x8762080, 0xc4209e1950, 0xc420a8be28, 0xab, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:491 +0xc9
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc420a8be00, 0x8762080, 0xc4209e1950, 0xc42089adb0, 0x5400, 0xc4200fa6f0, 0x87620c0, 0xc420938a80, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:419 +0x89b
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc4200fa5a0, 0x87620c0, 0xc420938a80, 0x5e81880, 0xc4205f4380, 0x0, 0x0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:415 +0x917
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func8.1(0x5e836e0, 0xc420a99040)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/server/server.go:536 +0x106
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4200b61b0, 0xc4203fe030, 0x5e836e0, 0xc420a99040, 0xc4205b03e0)
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
        /Users/jordan/repo/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x239
```

--- a/pkg/sql/index_selection.go
+++ b/pkg/sql/index_selection.go
@@ -1619,54 +1619,60 @@ func applyConstraintFlat(
 		return t
 	}
 
-	// More special casing: constraint says "a IN (x, y, z...)"
-	if cOp == parser.In {
-		ctuple := cdatum.(*parser.DTuple)
-		switch t.Operator {
-		case parser.Is, parser.EQ:
+	// Additional special casing.
+	switch t.Operator {
+	case parser.LE, parser.LT, parser.GE,
+		parser.GT, parser.NE, parser.IsNot:
+		if cOp == parser.In {
+			// The general case only handles range constraints.
+			// Stop here.
+
+			// Note: if we ever support other constraints than IN and the
+			// range operators, the condition on cOp here must be extended.
+			return t
+		}
+
+		// Otherwise, the general case below knows about all these
+		// cases. Go forward.
+
+	case parser.Is, parser.EQ:
+		// (Expr IS ...) / (Expr = ...)
+
+		if cOp == parser.In {
+			// constraint says "a IN (x, y, z...)"
 			// Expr asks "= Cst" or IS TRUE / IS FALSE: check whether the
 			// value is in the IN set. If it is not, we're sure it never
 			// matches. Otherwise we don't know.
+			ctuple := cdatum.(*parser.DTuple)
 			i := sort.Search(len(ctuple.D), func(i int) bool {
 				return ctuple.D[i].(parser.Datum).Compare(datum) >= 0
 			})
-			if i >= len(ctuple.D) || ctuple.D[i].Compare(datum) != 0 {
+			if i == len(ctuple.D) || ctuple.D[i].Compare(datum) != 0 {
 				return parser.DBoolFalse
 			}
+			return t
+		}
+		// Continue to the general case below.
 
+	case parser.In:
+		// Expr: "a IN (t, u, v, ...)"
+
+		switch cOp {
 		case parser.In:
-			// Expr: "a IN (t, u, v, ...)"
+			// constraint says "a IN (x, y, z...)"
 			// true if (x, y, z, ...) is a subset of (t, u, v, ...)
 			// unknown otherwise
 			if ttuple, ok := datum.(*parser.DTuple); ok {
 				// Note: A is a subset of B iff A \ B == empty.
+				ctuple := cdatum.(*parser.DTuple)
 				diff := ctuple.SortedDifference(ttuple)
 				if len(diff.D) == 0 {
 					return parser.DBoolTrue
 				}
 			}
 
-		case parser.NotIn:
-			// Expr: "a NOT IN (t, u, v, ...)"
-			// True if none of the values in (x, y, z...) are in (t, u, v, ...)
-			// unknown otherwise
-			if ttuple, ok := datum.(*parser.DTuple); ok {
-				// Note: A inter B is empty iff A \ B == A
-				diff := ctuple.SortedDifference(ttuple)
-				if len(diff.D) < len(ctuple.D) {
-					return parser.DBoolTrue
-				}
-			}
-		}
-
-		return t
-	}
-
-	// More special casing: constraint says [a = value]
-	// and expression is [a != ...], [a IN ...] or [a NOT IN ...].
-	if cOp == parser.EQ {
-		switch t.Operator {
-		case parser.In:
+		case parser.EQ:
+			// constraint says "a = Cst"
 			// true if the known value is in the IN set, false otherwise.
 			if tuple, ok := datum.(*parser.DTuple); ok {
 				i := sort.Search(len(tuple.D), func(i int) bool {
@@ -1677,8 +1683,31 @@ func applyConstraintFlat(
 				}
 			}
 			return parser.DBoolFalse
+		}
 
-		case parser.NotIn:
+		// The general case below doesn't know about
+		// IN expressions. Can't go further.
+		return t
+
+	case parser.NotIn:
+		// Expr: "a NOT IN (t, u, v, ...)"
+
+		switch cOp {
+		case parser.In:
+			// constraint says "a IN (x, y, z...)"
+			// True if none of the values in (x, y, z...) are in (t, u, v, ...)
+			// unknown otherwise
+			if ttuple, ok := datum.(*parser.DTuple); ok {
+				// Note: A inter B is empty iff A \ B == A
+				ctuple := cdatum.(*parser.DTuple)
+				diff := ctuple.SortedDifference(ttuple)
+				if len(diff.D) < len(ctuple.D) {
+					return parser.DBoolTrue
+				}
+			}
+
+		case parser.EQ:
+			// constraint says "a = Cst"
 			// false if the known value is in the NOT IN set, true otherwise.
 			if tuple, ok := datum.(*parser.DTuple); ok {
 				i := sort.Search(len(tuple.D), func(i int) bool {
@@ -1690,6 +1719,15 @@ func applyConstraintFlat(
 			}
 			return parser.DBoolTrue
 		}
+
+		// The general case below doesn't know about
+		// NOT IN expressions. Can't go further.
+		return t
+
+	default:
+		// We have a comparison expression with a comparator which
+		// is not handled below (IN, = ANY, etc). Can't go further.
+		return t
 	}
 
 	// General case: constraint and expression both specify a range.

--- a/pkg/sql/index_selection_test.go
+++ b/pkg/sql/index_selection_test.go
@@ -688,6 +688,16 @@ func TestApplyConstraints(t *testing.T) {
 		// elided during constraint analysis before constraint
 		// propagation.
 		{`a <= 1 AND COALESCE(a != 2, true)`, `a`, `COALESCE(true, true)`},
+		// Regression tests: #13707
+		// The following tests must achieve a constraint on `a` and an
+		// expression to simplify that contains `a` in a previously
+		// unhandled comparison operator. We use OR so that analyzeExpr
+		// doesn't decompose further.
+		{`a < 3 AND (b < 2 OR a IN (0,1,2))`, `a`, `(b < 2) OR (a IN (0, 1, 2))`},
+		{`a < 3 AND (b < 2 OR a NOT IN (0,1,2))`, `a`, `(b < 2) OR (a NOT IN (0, 1, 2))`},
+		{`a < 3 AND (b < 2 OR a = ANY ARRAY[0,1,2])`, `a`, `(b < 2) OR (a = ANY {0,1,2})`},
+		{`a IN (0, 2, 3) AND (b < 2 OR a <= 4)`, `a`, `(b < 2) OR (a <= 4)`},
+		{`a IN (0, 2, 3) AND (b < 2 OR a = 2)`, `a`, `(b < 2) OR (a = 2)`},
 	}
 	for _, d := range testData {
 		t.Run(d.expr+"~"+d.expected, func(t *testing.T) {

